### DISCUSSION TITLE
Editing family of async atom in memory

### DISCUSSION THREAD

**Discussion by stevemolitor** [pmndrs/jotai#1964] [Q&A] [answered]
I have an app where users edit load and edit complex JSON documents in memory. It's highly reactive - changing one value causes distant subscribers on a page to re-calculate and re-render their bits. The documents are loaded async from the server, and then edited in the browser.

What's the best way to handle this in Jotai? In Recoil I could manage this with a single atom or atom family. For example you could initialize an atom with a promise that fetches the value, or asynchronously call `setSelf`.  What's the most idiomatic way to do something similar in Jotai? It doesn't seem you can have a single atom that combines an async get with a synchronous set. 

Here is a [CodeSandbox](https://codesandbox.io/p/sandbox/github/stevemolitor/jotai-edit-server-state-in-browser/tree/main) with my current solution. A user can pick an "order" document from a dropdown, which loads the order from the server, and then the user edit the order in memory. The "discount" component reacts to order changes:

![](https://cdn.zappy.app/10c96efd11e5f2f4a36f7cfefa6bde3f.gif)

I use three atom families: 

  - `serverOrdersFamily` - family of immutable atoms with an async get to fetch orders from the server.
  - `clientOrdersFamily` - family of mutable atoms with user edits. Each atom is initially `undefined`.
  - `ordersFamily` - facade family that merges between the server and client orders. It merges on set. 

```typescript
// atom holding currently selected order id
const orderIdAtom = atom<OrderId>("1");

// atom family of orders fetched from server
const serverOrdersFamily = atomFamily((id: OrderId) =>
  atom(async () => fetchOrder(id))
);

// atom family of orders edited in browser
const clientOrdersFamily = atomFamily((_id: OrderId) =>
  atom<Order | undefined>(undefined)
);

// facade orders family that merges server and client edited order
const ordersFamily = atomFamily((id: OrderId) =>
  atom(
    (get) => {
      const clientOrder = get(clientOrdersFamily(id));
      if (clientOrder) {
        return clientOrder;
      }
      return get(serverOrdersFamily(id));
    },
    async (get, set, edit: Partial<Order>) => {
      const oldOrder = await get(ordersFamily(id));
      const newOrder = { ...oldOrder, ...edit };
      set(clientOrdersFamily(id), newOrder);
    }
  )
);
```

Is this the best or most idiomatic approach? 

There are only two order props in this example, but imagine a large, complex document with lots of nesting the the user can graphically edit, with lots of reactivity.

**Comment by dai-shi** [comment, marked as answer]
Your code looks good. Alternatively, I would probably use `atomWithDefault`.
